src/vss_tools/exporters/graphql.py

```diff
@@ -1,134 +1,558 @@
+        extra_args = {}
+
+        if allowed == "":
+            # No allowed values are specified, then use the scalars from datatype_map as the field type
+            field_type = datatype_map[child_leaf_metadata_row["datatype"]]
+
+        else:
+            # Allowed values are specified
+            allowed_enum = gql_allowed_enums[child_fqn]
+
+            datatype = child_leaf_metadata_row["datatype"]
+            if datatype and is_array(datatype):
+                field_type = graphene.List(allowed_enum)
+            else:
+                field_type = allowed_enum
+
+        if unit:
+            try:
+                quantity_kind = dynamic_units[unit].quantity  # Quantity kind name as described in VSS
+                enum_name = get_unit_enum_name(quantity_kind)  # GraphQL enum name for the quantity kind
+                unit_enum = gql_unit_enums[enum_name]
+                unit_value = dynamic_units[unit].unit
+
+                if unit_value is not None:
+                    unit_enum_value = get_gql_name(unit_value, GQLElementType.ENUM_VALUE)
+                else:
+                    raise ValueError(f"Unit value for '{unit}' is None")
+
+                if unit_enum_value not in unit_enum._meta.enum.__members__:
+                    raise ValueError(f"Unit enum value '{unit_enum_value}' not found in enum '{unit_enum}'")
+                else:
+                    extra_args["unit"] = graphene.Argument(type_=unit_enum, default_value=unit_enum_value)  # type: ignore
+
+                gql_fields[field_name] = Field(name=field_name, type_=field_type, description=field_description, args=extra_args)
+            except ValueError as e:
+                print(e)
+        else:
+            gql_fields[field_name] = Field(name=field_name, description=field_description, type_=field_type)
+
+        global mapping_leaves_df
+        mapping_leaves_df.loc[child_fqn, ["gql_field", "in_gql_type"]] = [field_name, gql_type_name]
+
+    # Create a field for each sub branch and call the creation of the GraphQL type recursively
+    for child_fqn, child_branch_metadata_row in child_branches.iterrows():
+        field_name = get_gql_name(child_branch_metadata_row["name"], GQLElementType.FIELD)
+        field_type = create_gql_object_type(child_fqn)
+        if child_fqn in branches_with_instances.index:
+            field_name += "_s"
+            field_type = graphene.List(field_type)
+        gql_fields[field_name] = Field(name=field_name, type_=field_type)
+
+    global mapping_branches_df
+    mapping_branches_df.loc[fqn, "gql_type"] = gql_type_name
+    return type(gql_type_name, (graphene.ObjectType,), gql_fields, description=gql_type_description)  # type: ignore
+
+
+def get_graphql_schema(tree: VSSNode) -> graphene.Schema:
+    """Create a GraphQL schema from the VSS tree."""
+    global vss_metadata_df, vss_branches_df, vss_leaves_df, gql_unit_enums, gql_allowed_enums, gql_instance_enums
```

Comment:
So many globals. Isn't it possible to pass them through where needed?

Reply:
It is possible, but as I have many functions, it will imply passing them over multiple times.